### DISCUSSION TITLE
ETIMEOUT using ntptime.py ESP32

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#10632] [Libraries & Drivers] [answered]
After reading this discussion:
https://github.com/orgs/micropython/discussions/10611
I "up-dated" my getUTC.py program taking the stated requirement
of the "15 second minimum polling interval" into account.

Even making this correction I was still getting ETIMEOUT errors, about
10% of time, when polling every 25 seconds.

By putting a try/except around every function call I ended up with errors with this call:
```py
    try:
        msg = s.recv(48)
        print (f'NTP response: {msg}')
        utime.sleep_ms(100)

    except OSError as e:
        try:
            with open('errors.txt', 'a') as outfile:
                outfile.write('did NOT recv the msg' + '\n')
                outfile.write('error = ' + str(e) + '\n')

            print (f'wait 60 seconds before doing a machine.reset()')
            utime.sleep_ms(100)

            machine.reset()
        except OSError:
            pass 
```
I am currently using `pool.ntp,org` but have tried `nz.pool.ntp.org`and I am not convinced that is any better.
I previously found that a 10 second timeout seemed to be more successful.

Any suggestions for testing and/or improving the success-rate are appreciated.

**Comment by peterhinch** [comment, marked as answer]
By default the [timeout](https://github.com/micropython/micropython-lib/blob/863a018b892f0bf901fa09844aa3428d5a5a00c1/micropython/net/ntptime/ntptime.py#L15) is set to 1s. It can be increased as per the code comment, but there is a question mark over the point of having a time signal that arrives after a long delay.

In my experience timeouts are fairly common. It just depends on your internet connection.

**Reply by davefes** [reply]
Somehow I missed your ntptime.py in your `micropython-samples-master`.

How do you adjust `hour` in `def time()`, manually?

**Reply by peterhinch** [reply]
If you mean `hrs_offset`, MP has no support for timezones. The caller needs to determine the value from the locale and any daylight savings in effect.

**Reply by davefes** [reply]
I call ntptime then apply the correction before settime.  Is there a way `to determine the value from the locale` on an ESP32 that re-boots "in the wild"?

**Reply by peterhinch** [reply]
Aside from a GPS module and a lookup table I don't know of one. When setting up a Linux box, you always have to tell it which country it is in.

**Comment by efahl** [comment]
If you want reasonably accurate network time on a uPython device, you need to either implement a full NTP client, or point the current implementation to some known-good "zero" delay source.  I run ntpd on my router, letting it do all the heavy lifting from four NTP sources, and then set ntptime host to the router.  (I have also used chronyd on one of my servers for the same purpose, but like the router-based solution better.)

```
import ntptime
ntptime.host = '10.1.1.1'  # Main router running ntpd.
ntptime.settime()  # Measured delays are never more that 1 ms.
```

**Reply by davefes** [reply]
Thanks for the responses.  Good point about timeout periods.  In my case(s) I don't need accurate time I just want a NTP time reliably, when I ask for it.

Appears either the sendto the NPT server or the return message fails.  Currently, I just bail-out and keep trying until I get a proper response.

@efahl I will do some reading-up on a local NTP server.

**Comment by peterhinch** [comment]
@efahl Your response about `ntpd` prompted me to experiment - the outcome might be of use to some here...

I have an ancient Raspberry Pi running 24/7. It runs standard Raspbian and serves as a Mosquitto broker. Without any configuration changes the following worked on a Pico W:
```python
ntptime.host="192.168.0.10"  # Address of the Pi
ntptime.time()
```
I proved that this really was working by invalidating the IP - the query then timed out.
So it seems that a bog standard Pi acts as a timeserver by default.

**Reply by efahl** [reply]
Hi Peter,

My Pi with plain raspian 32 and nothing else installed doesn't serve time, so I'll bet it's the Mosquitto code that's implementing a time server. 

```
efahl@ubuntu$ ntpdate -q rtr00
server fd0a:bad:dad::1, stratum 3, offset -0.000786, delay 0.02855
server 10.1.1.1, stratum 3, offset -0.000760, delay 0.02859
 1 Feb 07:35:43 ntpdate[3797903]: adjust time server fd0a:bad:dad::1 offset -0.000786 sec

efahl@ubuntu$ ntpdate -q pi03
 1 Feb 07:35:49 ntpdate[3797905]: no server suitable for synchronization found
```

On the router I see ``ntpd`` running, but on the Pi I see nothing on port 123.

```
root@rtr00$ netstat -tunlp | grep 123
udp        0      0 :::123                  :::*            4531/ntpd
```


Eric

**Reply by efahl** [reply]
After a quick ``sudo apt install -y ntp ntpdate`` on my Pi, it now is a time server, without any further config changes.  Should be pretty much the same on any debian-derivative Linux (and most other distros, I'd suspect).
```
efahl@pi03$ sudo systemctl status ntp.service
● ntp.service - Network Time Service
     Loaded: loaded (/lib/systemd/system/ntp.service; enabled; vendor preset: enabled)
     Active: active (running) since Wed 2023-02-01 07:56:37 PST; 2min 34s ago
       Docs: man:ntpd(8)
    Process: 4002 ExecStart=/usr/lib/ntp/ntp-systemd-wrapper (code=exited, status=0/SUCCESS)
   Main PID: 4011 (ntpd)
      Tasks: 2 (limit: 3720)
        CPU: 115ms
     CGroup: /system.slice/ntp.service
             └─4011 /usr/sbin/ntpd -p /var/run/ntpd.pid -g -u 118:127

efahl@pi03$  ntpdate -q localhost
server ::1, stratum 4, offset -0.000001, delay 0.02583
server 127.0.0.1, stratum 4, offset -0.000032, delay 0.02579
 1 Feb 08:00:19 ntpdate[4366]: adjust time server 127.0.0.1 offset -0.000032 sec
```

**Reply by peterhinch** [reply]
@efahl Thanks for that. It seems you're right. On my Pi:
```bash
root@raspberrypi:~# uptime -s
2022-06-18 01:48:58
root@raspberrypi:~# systemctl status ntp.service
● ntp.service - LSB: Start NTP daemon
   Loaded: loaded (/etc/init.d/ntp)
   Active: active (running) since Sat 2022-06-18 00:17:14 BST; 7 months 15 days ago
  Process: 457 ExecStart=/etc/init.d/ntp start (code=exited, status=0/SUCCESS)
   CGroup: /system.slice/ntp.service
           └─482 /usr/sbin/ntpd -p /var/run/ntpd.pid -g -u 107:112

Warning: Journal has been rotated since unit was started. Log output is incomplete or unavailable.
root@raspberrypi:~root@raspberrypi:~# uptime -s
2022-06-18 01:48:58
# systemctl status mosquitto.service
● mosquitto.service - LSB: mosquitto MQTT v3.1 message broker
   Loaded: loaded (/etc/init.d/mosquitto)
   Active: active (exited) since Sat 2022-06-18 00:17:12 BST; 7 months 15 days ago
  Process: 337 ExecStart=/etc/init.d/mosquitto start (code=exited, status=0/SUCCESS)

Warning: Journal has been rotated since unit was started. Log output is incomplete or unavailable.
```
So ntpd started just after mosquitto. I started mosquitto manually after a reboot resulting from a power outage.
